### PR TITLE
WIP - add zstd compression

### PR DESCRIPTION
**Pull request by darshanime**
### 1. What does this change do, exactly?
[WIP] - Add zstd compression

Currently, the new `zstd` package is a placeholder and just adds a new header. 
I realized that it will be very similar to the `gzip` package, with the same block in the Caddyfile etc. Instead of duplicating much of the code, would it be a better way to refactor compression as a general directive and have gzip, zstd (and something else too in the future) add just the appropriate `Writer`?

Comments, @abiosoft?

This can be done without breaking existing Caddyfiles. 

### 2. Please link to the relevant issues.
https://github.com/mholt/caddy/issues/2153

### 3. Which documentation changes (if any) need to be made because of this PR?
Plugin docs

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I am willing to help maintain this change if there are issues with it later

### CODE REVIEW

**Comment by darshanime** [comment]
> Instead of duplicating much of the code, would it be a better way to refactor compression as a general directive and have gzip, zstd (and something else too in the future) add just the appropriate Writer?

comments @abiosoft @mholt?

**Comment by abiosoft** [comment]
> > Instead of duplicating much of the code, would it be a better way to refactor compression as a general directive and have gzip, zstd (and something else too in the future) add just the appropriate Writer?
> 
> comments @abiosoft @mholt?

I'm in favour of this as well, the code is easily reusable. Maybe we can consider depreciating `gzip` and introducing a `compress` directive that takes `gzip`, `brotli`, `zstd` e.t.c. as config param. 

I think the main issue would be if the compression is provided by a plugin. But in this case `zstd` is gonna be part of standard Caddy plugins.

**Comment by tobya** [comment]
I think we would need to do this before 1.0 and slowly deprecate the `gzip` directive.

If @mholt like the idea would you be willing to work on a `compress` directive @darshanime ?

**Comment by mholt** [comment]
Yes, now's a great time to do it. Let's have `compress` (and keep `gzip` for a while so we can deprecate it gracefully) and find a way to put all the compression code into there.

Note that brotli compression is only supported for pre-compressed, static assets (mainly because brotli is expensive to compress, and no good pure-Go encoders exist last time I checked).

**Comment by darshanime** [comment]
yeah sure @tobya, I'll love to work on this. 
On it! 👍

**Comment by tobya** [comment]
Could we open an issue for this so we can discuss as an issue rather than on a PR

**Comment by darshanime** [comment]
@tobya sure. I've raised https://github.com/mholt/caddy/issues/2173

File: caddyhttp/compress/zstd.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// limitations under the License.
+
+package compress
+
+import (
+	"io/ioutil"
+	"io"
+	"sync"
+
+	"github.com/DataDog/zstd"
```
Comment (elcore) [inline caddyhttp/compress/zstd.go]: This library is using `cgo`, I don't like it...

99,8 % of the code is C!!!!!!

Comment (elcore) [reply caddyhttp/compress/zstd.go]: There will be cross compile issues!!!!

Comment (mholt) [inline caddyhttp/compress/zstd.go]: Unfortunately, cgo is an unacceptable dependency for Caddy core. You'll have to find a package that doesn't use C; it has to be a pure Go implementation.

Comment (elcore) [reply caddyhttp/compress/zstd.go]: @mholt I had the same thought (see above)

**Comment by darshanime** [comment]
There are only 2 official bindings (http://facebook.github.io/zstd/#other-languages) and both use cgo. We can defer `zstd` until someone implements it in pure go.
Are we merging the `compress` directive change or that is redundant now?

**Comment by mholt** [comment]
I don't think there's much of a need for it yet; let's just wait until there's a Go implementation of another compression standard that is fit for a web server. Thanks!

**Comment by tobya** [comment]
@darshanime Thanks for all your work on this.

**Comment by mholt** [comment]
Zstd compression is implemented in Caddy 2 (see the v2 branch).

**Pull request closed without merging**